### PR TITLE
Replace `secp256k1` keys with `util::key` keys in BIP32

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

File: src/util/bip32.rs [hunk 1]
```diff
@@ -425,7 +431,7 @@ impl ExtendedPrivKey {
         match i {
             ChildNumber::Normal {..} => {
                 // Non-hardened key: compute public data and use that
-                hmac_engine.input(&PublicKey::from_secret_key(secp, &self.secret_key).serialize()[..]);
+                PublicKey::from_private_key(secp, &self.secret_key).write_into(&mut hmac_engine);
```
Comment (dongcarl) [inline src/util/bip32.rs]: Must be compressed serialization.

Comment (apoelstra) [reply src/util/bip32.rs]: fixed

File: src/util/bip32.rs [hunk 2]
```diff
@@ -24,10 +24,10 @@ use std::str::FromStr;
 
 use byteorder::{BigEndian, ByteOrder, ReadBytesExt};
 use bitcoin_hashes::{hash160, sha512, Hash, HashEngine, Hmac, HmacEngine};
-use secp256k1::key::{PublicKey, SecretKey};
 use secp256k1::{self, Secp256k1};
 
 use network::constants::Network;
+use util::key::{PublicKey, PrivateKey};
```
Comment (stevenroose) [inline src/util/bip32.rs]: nit: `util::key` sorts below `util::base58`

File: src/util/key.rs [hunk 3]
```diff
@@ -152,6 +152,13 @@ impl FromStr for PrivateKey {
     }
 }
 
+impl ops::Index<ops::RangeFull> for PrivateKey {
```
Comment (stevenroose) [inline src/util/key.rs]: :+1:

File: src/util/key.rs [hunk 4]
```diff
@@ -35,7 +35,7 @@ pub struct PublicKey {
 
 impl PublicKey {
     /// Write the public key into a writer
-    pub fn write_into<W: io::Write>(&self, writer: &mut W) {
+    pub fn write_into<W: io::Write>(&self, mut writer: W) {
```
Comment (stevenroose) [inline src/util/key.rs]: What does this change exactly? I never saw this to be honest..

Comment (apoelstra) [reply src/util/key.rs]: It lets you pass a `&mut [u8]` to the function.

Comment (apoelstra) [reply src/util/key.rs]: See also https://rust-lang-nursery.github.io/api-guidelines/interoperability.html#generic-readerwriter-functions-take-r-read-and-w-write-by-value-c-rw-value

File: src/util/bip32.rs [hunk 5]
```diff
@@ -57,8 +57,8 @@ pub struct ExtendedPrivKey {
     pub parent_fingerprint: Fingerprint,
     /// Child number of the key used to derive from parent (0 for master)
     pub child_number: ChildNumber,
-    /// Secret key
-    pub secret_key: SecretKey,
+    /// Private key
+    pub secret_key: PrivateKey,
```
Comment (dongcarl) [inline src/util/bip32.rs]: Perhaps we can rename this to `private_key`?

Comment (apoelstra) [reply src/util/bip32.rs]: Sure. I considered this but wanted to minimize the diff. But I guess we'll be tripping over this for years if I don't change it ..

File: src/util/bip32.rs (hunk 2 above)
Comment (apoelstra) [reply src/util/bip32.rs]: Fixed.

**Pull request merged**